crates/backend/src/codegen.rs

```diff
@@ -1496,3 +1511,91 @@ fn respan(input: TokenStream, span: &dyn ToTokens) -> TokenStream {
+            debug_assert!(func.is_function());
+            let iter = func.unchecked_into::<js_sys::Function>()
+                .call0(self)
+                .unwrap_throw();
+            // TODO debug check iter can be used as an iterator (e.g. check for `next` method)
+            iter.unchecked_into::<js_sys::Iterator>()
+        }
+    };
+    let iter_fn = match iterable {
+        ast::Iterable::MapLike => {
+            quote! {
+                /// Wraps the javascript object at `Symbol.iterator`, and presents it as a rust
+                /// `Iterator`. See ["The iterable protocol" at MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#The_iterable_protocol)
+                pub fn iter(&self)
+                    -> impl std::iter::Iterator<Item=(wasm_bindgen::JsValue, wasm_bindgen::JsValue)>
+                {
+                    use wasm_bindgen::{UnwrapThrowExt, JsCast};
+                    struct Iter(js_sys::Iterator);
+
+                    impl Iterator for Iter {
+                        type Item = (wasm_bindgen::JsValue, wasm_bindgen::JsValue);
+                        fn next(&mut self) -> Option<Self::Item> {
+                            let next = self.0.next().unwrap_throw();
+                            if next.done() {
+                                return None; // We discard the end value here.
+                            }
+                            let entry = next.value();
+                            debug_assert!(entry.has_type::<js_sys::Array>());
+                            let entry: js_sys::Array = entry.unchecked_into();
+                            let key = entry.get(0);
+                            let value = entry.get(1);
+                            Some((key, value))
+                        }
+                    }
+
+                    Iter(#get_iterator)
+                }
+            }
+        }
+        ast::Iterable::ArrayLike => {
+            quote! {
+                pub fn iter(&self) -> impl std::iter::Iterator<Item=wasm_bindgen::JsValue> {
+                    use wasm_bindgen::{UnwrapThrowExt, JsCast};
+                    struct Iter(js_sys::Iterator);
+
+                    impl Iterator for Iter {
+                        type Item = wasm_bindgen::JsValue;
+                        fn next(&mut self) -> Option<Self::Item> {
+                            let next = self.0.next().unwrap_throw();
+                            if next.done() {
+                                return None; // We discard the end value here.
+                            }
+                            Some(next.value())
+                        }
+                    }
+
+                    Iter(#get_iterator)
+                }
+            }
+        }
+        ast::Iterable::No => {
+            quote! {}
+        }
+    };
+    // if we have something to output and rust_name was given, then wrap, otherwise don't
+    if let Some(rust_name) = rust_name {
```

Comment:
The idea with this code is that either you wrap it in an `impl` yourself and pass `None`, or pass `Some(ident)` and and the function will wrap it in an `impl #ident`.

Reply:
An alternative to what I'm doing currently would be to put the `iter` method in a separate `impl` block, then both cases would be the `Some` branch and could be unified. The downside is the generated docs are slightly less nice to read (since they have more impl blocks). Would you like me to do this?